File: 0040-combination-sum-ii/0040-combination-sum-ii.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
    candidates.sort((a, b) => a - b)
    let result = [];
    
    let iterate = (index, sum, temp) =>{
        if(sum > target) return;
        
        if(sum == target){
            result.push(temp)
            return;
        }
        
        for(let i = index; i < candidates.length; i++){
            if(i != index && candidates[i] == candidates[i - 1]) continue;
            
            iterate(i + 1, sum + candidates[i], [...temp, candidates[i]])
        }
    }
    iterate(0, 0, [])
    return result;
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> result = new ArrayList<>();

        dfs(candidates, target, 0, new ArrayList<Integer>(), result);
        return result;
    }

    private void dfs(int[] candidates, int target, int start, List<Integer> comb, List<List<Integer>> result) {
        if (target < 0) {
            return;
        }

        if (target == 0) {
            result.add(new ArrayList<Integer>(comb));
            return;
        }

        for (int i = start; i < candidates.length; i++) {
            if (i > start && candidates[i] == candidates[i-1]) {
                continue;
            }

            if (candidates[i] > target) {
                break;
            }

            comb.add(candidates[i]);
            dfs(candidates, target - candidates[i], i + 1, comb, result);
            comb.remove(comb.size() - 1);
        }
    }
}